File: backend/src/models/Patient.js
import mongoose from 'mongoose';
import bcrypt from 'bcryptjs';

// Generate unique alphanumeric ID for patients
const generateUniqueId = () => {
  const prefix = 'PAT';
  const randomPart = Math.random().toString(36).substr(2, 6).toUpperCase();
  return `${prefix}${randomPart}`;
};

const patientSchema = new mongoose.Schema({
  userId: { type: String, unique: true, required: false },
  name: { type: String, required: true },
  email: { type: String, required: true, unique: true },
  password: { type: String, required: true },
  role: { type: String, default: 'patient' },
  
  // Patient-specific fields
  dateOfBirth: { type: Date, required: true },
  age: { type: Number, required: false },
  gender: { type: String, required: false, enum: ['male', 'female', 'other'] },
  phoneNumber: { type: String, required: true },
  emergencyContact: {
    name: String,
    phone: String,
    relationship: String
  },
  medicalHistory: [{
    condition: String,
    diagnosisDate: Date,
    status: String
  }],
  allergies: [String],
  currentMedications: [{
    name: String,
    dosage: String,
    frequency: String,
    duration: String,
    instructions: String,
    timing: [String], // ['morning', 'afternoon', 'night']
    foodTiming: String, // 'Before', 'After', 'With'
    prescribedBy: String,
    prescribedDate: String,
    prescriptionId: { type: mongoose.Schema.Types.ObjectId, ref: 'Prescription' },
    scheduleExplanation: String, // Explanation of how the schedule was generated
    smartScheduled: { type: Boolean, default: false }, // Whether schedule was auto-generated
    adherence: [{
      timestamp: { type: Date, default: Date.now },
      taken: { type: Boolean, required: true },
      notes: String,
      recordedBy: String
    }],
    lastEditedBy: String,
    lastEditedAt: Date
  }],
  
  // Visit history
  visits: [{
    visitDate: { type: Date, default: Date.now },
    visitType: { type: String, enum: ['medicine_prescription', 'consultation', 'follow_up', 'emergency'], default: 'medicine_prescription' },
    doctorId: { type: mongoose.Schema.Types.ObjectId, ref: 'Doctor' },
    doctorName: String,
    diagnosis: String,
    notes: String,
    medicines: [{
      name: String,
      dosage: String,
      frequency: String,
      duration: String,
      instructions: String
    }],
    followUpDate: Date,
    followUpRequired: { type: Boolean, default: false },
    createdAt: { type: Date, default: Date.now }
  }],
  
  // Caretaker assignments
  selectedCaretaker: {
    caretakerId: { type: mongoose.Schema.Types.ObjectId, ref: 'Caretaker' },
    caretakerUserId: { type: String }, // For easy lookup
    caretakerName: { type: String }, // Caretaker's name for display
    caretakerEmail: { type: String }, // Caretaker's email for display
    assignedAt: { type: Date, default: Date.now }
  },
  
  // Caretaker approvals
  caretakerApprovals: [{
    caretakerId: { type: mongoose.Schema.Types.ObjectId, ref: 'Caretaker' },
    status: { type: String, enum: ['pending', 'approved', 'rejected'], default: 'pending' },
    requestedAt: { type: Date, default: Date.now },
    approvedAt: Date,
    rejectedAt: Date
  }],
  
  // System fields
  isActive: { type: Boolean, default: true },
  lastLogin: Date,
  createdAt: { type: Date, default: Date.now },
  updatedAt: { type: Date, default: Date.now }
});

// Pre-save hook to generate unique ID, hash password, and calculate age
patientSchema.pre('save', async function(next) {
  try {
    // Always generate unique ID if not already set
    if (!this.userId) {
      this.userId = generateUniqueId();
    }
    
    // Hash password if modified
    if (this.isModified('password')) {
      this.password = await bcrypt.hash(this.password, 10);
    }
    
    // Calculate age from date of birth if not already set or if date of birth changed
    if (this.dateOfBirth && (!this.age || this.isModified('dateOfBirth'))) {
      const today = new Date();
      const birthDate = new Date(this.dateOfBirth);
      let age = today.getFullYear() - birthDate.getFullYear();
      const monthDiff = today.getMonth() - birthDate.getMonth();
      
      if (monthDiff < 0 || (monthDiff === 0 && today.getDate() < birthDate.getDate())) {
        age--;
      }
      
      this.age = age;
    }
    
    // Update timestamp
    this.updatedAt = new Date();
    
    next();
  } catch (error) {
    next(error);
  }
});

// Post-save hook to ensure userId is set
patientSchema.post('save', function(doc) {
  if (!doc.userId) {
    doc.userId = generateUniqueId();
    doc.save().catch(err => console.error('Error setting userId:', err));
  }
});

// Method to compare password
patientSchema.methods.comparePassword = async function(candidatePassword) {
  return bcrypt.compare(candidatePassword, this.password);
};

export default mongoose.model('Patient', patientSchema);

File: android-app/app/src/main/java/com/medalert/patient/data/model/Patient.kt
package com.medalert.patient.data.model

import android.os.Parcelable
import kotlinx.parcelize.Parcelize
import java.util.Date

@Parcelize
data class Patient(
    val _id: String = "",
    val id: String = "", // For compatibility with backend response
    val userId: String = "",
    val name: String = "",
    val email: String = "",
    val dateOfBirth: String = "",
    val age: Int = 0,
    val gender: String = "",
    val phoneNumber: String = "",
    val emergencyContact: EmergencyContact? = null,
    val medicalHistory: List<MedicalCondition> = emptyList(),
    val allergies: List<String> = emptyList(),
    val currentMedications: List<Medication> = emptyList(),
    val visits: List<Visit> = emptyList(),
    val selectedCaretaker: SelectedCaretaker? = null,
    val caretakerApprovals: List<CaretakerApproval> = emptyList(),
    val isActive: Boolean = true,
    val createdAt: String = "",
    val updatedAt: String = ""
) : Parcelable {
    // Helper function to get the correct ID
    fun getPatientId(): String {
        return if (_id.isNotEmpty()) _id else id
    }
    
    // Helper function to get the user-friendly patient ID
    fun getUserFriendlyId(): String {
        return if (userId.isNotEmpty()) userId else getPatientId()
    }
    
    // Helper function to get display ID (user-friendly ID for UI)
    fun getDisplayId(): String {
        return if (userId.isNotEmpty()) {
            "Patient ID: $userId"
        } else {
            "Patient ID: ${getPatientId()}"
        }
    }
}

@Parcelize
data class EmergencyContact(
    val name: String = "",
    val phone: String = "",
    val relationship: String = ""
) : Parcelable

@Parcelize
data class MedicalCondition(
    val condition: String = "",
    val diagnosisDate: String = "",
    val status: String = ""
) : Parcelable

@Parcelize
data class Medication(
    val _id: String = "",
    val name: String = "",
    val dosage: String = "",
    val frequency: String = "",
    val duration: String = "",
    val instructions: String = "",
    val timing: List<String> = emptyList(),
    val foodTiming: String = "",
    val prescribedBy: String = "",
    val prescribedDate: String = "",
    val prescriptionId: String = "",
    val scheduleExplanation: String = "",
    val smartScheduled: Boolean = false,
    val adherence: List<AdherenceRecord> = emptyList(),
    val lastTaken: String = "",
    val updatedAt: String = "",
    val updatedBy: String = "",
    // Enhanced scheduling fields
    val startDate: String = "",
    val endDate: String = "",
    val totalTablets: Int = 0,
    val remainingTablets: Int = 0,
    val isActive: Boolean = true,
    val customSchedule: List<ScheduleEntry> = emptyList(),
    // Detailed timing and dose tracking
    val scheduledDoses: List<ScheduledDose> = emptyList(),
    val doseRecords: List<DoseRecord> = emptyList()
) : Parcelable

@Parcelize
data class AdherenceRecord(
    val timestamp: String = "",
    val taken: Boolean = false,
    val notes: String = "",
    val recordedBy: String = ""
) : Parcelable

@Parcelize
data class ScheduleEntry(
    val date: String = "", // YYYY-MM-DD format
    val time: String = "", // HH:MM format
    val label: String = "", // Morning, Afternoon, Evening, Night, Custom
    val isActive: Boolean = true,
    val tabletCount: Int = 1,
    val notes: String = ""
) : Parcelable

@Parcelize
data class Visit(
    val visitDate: String = "",
    val visitType: String = "",
    val doctorId: String = "",
    val doctorName: String = "",
    val diagnosis: String = "",
    val notes: String = "",
    val medicines: List<PrescribedMedicine> = emptyList(),
    val followUpDate: String = "",
    val followUpRequired: Boolean = false,
    val createdAt: String = ""
) : Parcelable

@Parcelize
data class PrescribedMedicine(
    val name: String = "",
    val dosage: String = "",
    val frequency: String = "",
    val duration: String = "",
    val instructions: String = ""
) : Parcelable

@Parcelize
data class SelectedCaretaker(
    val caretakerId: String = "",
    val caretakerUserId: String = "",
    val caretakerName: String = "",
    val caretakerEmail: String = "",
    val assignedAt: String = ""
) : Parcelable

@Parcelize
data class CaretakerApproval(
    val caretakerId: String = "",
    val status: String = "", // pending, approved, rejected
    val requestedAt: String = "",
    val approvedAt: String = "",
    val rejectedAt: String = ""
) : Parcelable

@Parcelize
data class ScheduledDose(
    val id: String = "",
    val time: String = "", // HH:MM format
    val label: String = "", // Morning, Afternoon, Evening, Night, Custom
    val dosage: String = "",
    val isActive: Boolean = true,
    val daysOfWeek: List<Int> = emptyList(), // 0=Sunday, 1=Monday, etc.
    val startDate: String = "",
    val endDate: String = "",
    val notes: String = ""
) : Parcelable

@Parcelize
data class DoseRecord(
    val id: String = "",
    val scheduledDoseId: String = "",
    val scheduledTime: String = "", // HH:MM format
    val scheduledDate: String = "", // YYYY-MM-DD format
    val actualTime: String = "", // HH:MM format when actually taken
    val actualDate: String = "", // YYYY-MM-DD format when actually taken
    val status: DoseStatus = DoseStatus.PENDING,
    val notes: String = "",
    val recordedAt: String = "",
    val recordedBy: String = "patient" // patient, caretaker, system
) : Parcelable

enum class DoseStatus {
    PENDING,    // Not yet taken
    TAKEN,      // Taken on time or within grace period
    MISSED,     // Not taken within grace period
    SKIPPED,    // Intentionally skipped by user
    LATE        // Taken but late
}